### PR TITLE
Add support for loading shell initialization file directly

### PR DESCRIPTION
**Pull request by iliajie**
This is a new PR to make it work in a more abstract way. I think you will like it more than the [previous one](https://github.com/webmin/webmin/pull/1785).

image

### CODE REVIEW

**Comment by iliajie** [comment]
Finished!

File: xterm/shellserver.pl [hunk 1]
```diff
@@ -28,70 +28,53 @@
...
+		$rcfile = "$uinfo[7]/$rcfile";
+		}
+	if (-r $rcfile) {
+		# Bash
+		if ($shellname eq 'bash') {
+			$shellexec = "$shellcmd --rcfile $rcfile";
+			}
+		# Zsh
+		elsif ($shellname eq 'zsh') {
+			$shellexec = "export ZDOTDIR=$rcdir ; $shellexec";
```
Comment (jcameron) [inline xterm/shellserver.pl:60]: Just do `$ENV{'ZDOTDIR'} = $rcdir;`

File: xterm/shellserver.pl [hunk 2]
```diff
@@ -28,70 +28,53 @@
...
+
+		# Cannot use login shell while passing other parameters,
+		# and it is not necessary as we already add init files
+		$shelllogin = undef;
+		}
 	}
-my ($shellfh, $pid) = &proc::pty_process_exec($shellcmd, $uid, $gid, $shellbin);
+my ($shellfh, $pid) = &proc::pty_process_exec($shellexec, $uid, $gid, $shelllogin);
 &reset_environment();
+my $shcmd = "'$shellexec".($shelllogin ? " $shelllogin" : undef)."'";
```
Comment (jcameron) [inline xterm/shellserver.pl:70]: Why do we need single quotes around the command?

File: xterm/shellserver.pl (hunk 1 above)
Comment (iliajie) [reply xterm/shellserver.pl]: Done.

**Comment by iliajie** [comment]
https://github.com/virtualmin/Virtualmin-Config/commit/061b5c7f32b98a0475c737a5214d7c8713c8e1ce

**Pull request merged**